contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -616,12 +592,28 @@ void USDScene::writeTransform( const Data *transform, double time )
 	}
 }
 
+namespace
+{
+
+const IECore::InternedString g_purposeAttributeName( "usd:purpose" );
+
+} // namespace
+
 bool USDScene::hasAttribute( const SceneInterface::Name &name ) const
 {
+	if( m_location->prim.IsPseudoRoot() )
+	{
+		// Can't store attributes here.
+		return false;
+	}
 	if( name == SceneInterface::visibilityName )
 	{
 		return pxr::UsdGeomImageable( m_location->prim ).GetVisibilityAttr().HasAuthoredValue();
 	}
+	else if( name == g_purposeAttributeName )
+	{
+		return pxr::UsdGeomImageable( m_location->prim ).GetPurposeAttr().HasAuthoredValue();
```

Comment:
I've lost track of whether we're now traversing non-imagable prims, but does this return the right thing in line with [this example](https://graphics.pixar.com/usd/docs/api/usd_geom_page_front.html#UsdGeom_ImageablePurpose)?

Reply:
Good question. We _do_ traverse non-imageable prims, so this is something we need to consider. Even though we're accessing via `UsdGeomImageable` and the prim isn't imageable, the current code appears to be loading `/Root.purpose` successfully. This means that we get different results from the example, which states that purpose on non-imageable prims should be ignored.

The only option we have for ignoring it is to not load it, but this does mean losing some information and some loss of round-tripping. But we couldn't round-trip the example in any case because we have lost the original prim type on loading, and will end up writing an Xform anyway. I do not want to get into _that_ particular topic right now. I suppose our main motivation for loading purpose is to allow users to replicate the USD visibility rules in Gaffer, so the most useful thing we can do in the short term is discard the purpose from non-imageable prims?

Reply:
Yeah, that sounds reasonable. Is there anything in [here](https://graphics.pixar.com/usd/docs/api/struct_usd_geom_imageable_1_1_purpose_info.html#a63a785994415244e9a44f0c4b2acffbf) that would make working out when to discard easier?

Reply:
Oh, that does look handy! Thanks, I hadn't even seen that. Let me see what I can do...

Reply:
I've pushed a new version which ignores purpose on non-imageable prims, including a test case based on the USD example. That `PurposeInfo` stuff turned out not to be useful, since the only public methods for getting it all compute the full inherited version, not the local version that we want. I did check the [internal implementation](https://github.com/PixarAnimationStudios/USD/blob/release/pxr/usd/usdGeom/imageable.cpp#L391-L405) though to verify that we're doing the expected thing.